tests/lib/rules/no-useless-escape.js

```diff
@@ -127,9 +127,59 @@ ruleTester.run("no-useless-escape", rule, {
     ],
 
     invalid: [
-        { code: "var foo = /\\#/;", errors: [{ line: 1, column: 12, endColumn: 13, message: "Unnecessary escape character: \\#.", type: "Literal" }] },
-        { code: "var foo = /\\;/;", errors: [{ line: 1, column: 12, endColumn: 13, message: "Unnecessary escape character: \\;.", type: "Literal" }] },
-        { code: "var foo = \"\\'\";", errors: [{ line: 1, column: 12, endColumn: 13, message: "Unnecessary escape character: \\'.", type: "Literal" }] },
+        {
+            code: "var foo = /\\#/;",
+            errors: [{
+                line: 1,
+                column: 12,
+                endColumn: 13,
+                message: "Unnecessary escape character: \\#.",
+                type: "Literal",
+                suggestions: [{
+                    desc: "Remove unnecessary escape.",
+                    fix: { range: [11, 12], text: "" }
+                }, {
+                    desc: "Escape backslash to include it in the RegExp.",
+                    fix: { range: [11, 11], text: "\\" }
+                }]
+            }]
+        },
+        {
+            code: "var foo = /\\;/;",
+            errors: [{
+                line: 1,
+                column: 12,
+                endColumn: 13,
+                message: "Unnecessary escape character: \\;.",
+                type: "Literal",
+                suggestions: [{
+                    messageId: "removeEscape",
+                    output: "var foo = /;/;"
```

Comment:
I pushed up this work-in-progress to get some feedback on the approach here. One option for these tests is to do something like what I have in the above test, where we assert against the suggestion fix data e.g.:
```js
fix: { range: [11, 12], text: "" }
```

One problem with this is that it is hard to look at and verify that it will work properly.

An alternative approach that I was exploring was to use something like this `output` key where we define the result of applying the suggestion on the original code. This seems a lot easier to write good tests with, but it doesn't actually exist on the suggestion data so it could potentially be confusing. Thoughts?

Reply:
Hmm.. That's a good question. I do think output approach is easier to deal with, but we don't have any way of getting that data (unless we run suggestions through auto-fix code inside tester)... So I think we'll have to stick with the ranges at least for now. We can change it later, if we find better approach.

Reply:
I'd really like to make sure the user experience is good here-- otherwise, I'm worried nobody would use the feature. So I'm in favor of getting the output to work (and having RuleTester run the code through the autofixer as needed). It seems like this shouldn't cause regressions because it would only happen when someone adds a suggestion assertion object to the test case.

Let me know if I'm missing something.

Reply:
Alright, I've added support for both options. See the [Testing Suggestions section](https://github.com/eslint/eslint/pull/12384/files#diff-10f97728420b529b7899f7f778241b95R899) in `docs/developer-guide/nodejs-api.md` for docs.

I'm running the `SourceCodeFixer.applyFixes` when a suggestion `output` is specified, which helped me catch a bug in my original implementation of suggestions for the `no-useless-escape` rule which I definitely would have missed without this.